### PR TITLE
Solutions to Lecture 3

### PR DESCRIPTION
**Pull request by duobei**
### Solutions for Lecture 3

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture3.hs [hunk 1]
```diff
@@ -49,15 +48,24 @@ data Weekday
...
-toShortString = error "TODO"
+
+toShortString :: Weekday -> String
+toShortString Monday = "Mon"
+toShortString Tuesday = "Tue"
+toShortString Wednesday = "Wed"
+toShortString Thursday = "Thu"
+toShortString Friday = "Fri"
+toShortString Saturday = "Sat"
+toShortString Sunday = "Sun"
```
Comment (chshersh) [inline src/Lecture3.hs]: That's the correct implementation 👍🏻 
But can you implement it without pattern matching on each constructor? 🤔

File: src/Lecture3.hs [hunk 2]
```diff
@@ -210,8 +246,23 @@ types that can have such an instance.
...
+
+  foldr f z (List1 x xs) = f x (foldr f z xs)
+
+
+instance Foldable Treasure where
+  foldMap = helper mempty
+    where
+      helper :: p -> (t -> p) -> Treasure t -> p
+      helper x _ NoTreasure = x
+      helper _ f (SomeTreasure x) = f x
```
Comment (chshersh) [inline src/Lecture3.hs]: I think you don't need a `helper` function here and can implement the same logic on the top-level for `foldMap` directly

**Review by chshersh — APPROVED** [review]
Looks correct! Feel free to merge this PR, it looks already good 🙂

File: src/Lecture3.hs (hunk 2 above)
Comment (duobei) [reply src/Lecture3.hs]: yup

File: src/Lecture3.hs (hunk 1 above)
Comment (duobei) [reply src/Lecture3.hs]: Nice. As `Weekday` has derived `Show`, I could use another way.

**Pull request merged**